README.md

[![Arduino CI](https://github.com/RobTillaart/bitHelpers/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/bitHelpers/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/bitHelpers/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/bitHelpers/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/bitHelpers/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/bitHelpers/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/bitHelpers.svg?maxAge=3600)](https://github.com/RobTillaart/bitHelpers/releases)


# bitHelpers

Arduino library with functions on bit level.


## Description

This library contains functions to manipulate bits and bit patterns in an 
efficient way. 
For most functions a 8 - 64 bit optimized version exist. 

The library is tested on ESP32 and Arduino UNO but not for all possible values. 
other platforms are expected to work without modification. 
If they don't please file an issue on GitHub.

New bit functions can be added or investigated, please file an issue on GitHub.


## Interface


### 0.1.0

BitCount, several implementations to compare performance.

- **uint8_t bitCountReference(uint32_t value)** returns number of bits set in a value.
- **uint8_t bitCountKR(uint32_t value)** Kerningham Ritchie bitCount.
- **uint8_t bitCountArray(uint32_t value)** count per nybble with lookup table.
- **uint8_t bitCountF1(uint32_t value)** SWAG algorithm variant.
- **uint8_t bitCountF2(uint32_t value)** SWAG algorithm variant.

BitCount - fastest version, SWAG algorithm

- **uint8_t  bitCount(uint8_t value)** available for 16, 32 and 64 bit.

Reverse: uint8_t .. uint64_t

- **T bitReverse(T value)** reverses bits in a uint8_t .. uint64_t.
- **T nybbleReverse(T value)** reverses nibbles (4 bit) in a uint8_t .. uint64_t.
- **T byteReverse(T value)** reverses bytes (8 bit) in a uint16_t .. uint64_t.
- **T wordReverse(T value)** reverses words (16 bit) in uint32_t and uint64_t.

Swap upper and lower half: uint8_t .. uint64_t.

- **T swap(T value)** 0x12345678 ==> 0x56781234.

Rotate Left / Right: uint8_t .. uint64_t
if pos larger than # bits original value is returned.

- **T bitRotateLeft(T value, uint8_t pos)**
- **T bitRotateRight(T value, uint8_t pos)** 

BitFlip: uint8_t .. uint64_t  a.k.a toggle
if pos larger than # bits original value is returned.

- **T bitFlip(T value, uint8_t pos)** flips a single bit at pos

BitRot: uint8_t .. uint64_t

- **T bitRot(T value, float chance = 0.5)** random damage to a single bit of a value,
chance = float 0.0 .. 1.0 that one random bit is toggled. 
**bitRot()** is a function that can be used to mimic single bit errors in communication protocols.  
*Note: a chance of 50% for 2 uint8_t is not equal to 50% chance for 1 uint16_t.*


### 0.1.1 added

How many bits are needed to store / transmit a number?

- **bitsNeededReference(n)** reference implementation for uint8_t to uint64_t.
- **bitsNeeded(n)** A 'recursive strategy' for uint8_t .. uint64_t provides a fast answer. 

The following functions are made as the normal **bitset()** etcetera do not work for 64 bit.
These functions are optimized for speed for **AVR**, **ESP32** and **ESP8266**. 

- **void bitSet64(uint64_t & x, uint8_t bit)** set bit of uint64_t
- **void bitClear64(uint64_t & x, uint8_t bit)** clear bit of uint64_t
- **void bitToggle64(uint64_t & x, uint8_t bit)** toggle bit of uint64_t
- **void bitWrite64(uint64_t & x, uint8_t bit, uint8_t value)** set bit of uint64_t to 0 or 1
- **void bitRead64(uint64_t & x, uint8_t bit)** reads bit from uint64_t 

Also added are macro versions of these five functions.

- **mbitSet64(x, bit)** set bit of uint64_t
- **mbitClear64(x, bit)** clear bit of uint64_t
- **mbitToggle64(x, bit)** toggle bit of uint64_t
- **mbitWrite64(x, bit, value)** set bit of uint64_t to 0 or 1
- **mbitRead64(x, bit)** reads bit from uint64_t 


### 0.1.2 and beyond

See CHANGELOG.md


## BitReverse n bit number

Trick to reverse a number of n bits  ( 0 < n < 32 ).
Could also be done similar with 64 bit and or byte / nibble reverse.

not as fast as a dedicated version.
```cpp
uint32_t bitReverse(uint32_t x, uint8_t n)
{
  uint32_t r = bitReverse(x);
  return r >> (32 - n);
}
```
Could be added in next release...


## Future

- improve documentation
- improve readability of code
- add performance tests


#### Functions add

- besides **bitRot()** one can also have timing issues when clocking in bits. 
A function could be created to mimic such timing error, by shifting bits from a 
specific position. e.g. 
- **parShiftLeft(00001010, 3)** ==> 00011010
- **bitBurst(00000000, 3)** ==>  00111000 any group of 3 bits will toggle. edges?
- **bitNoggle(value, bit)** - toggle all but one bit. (why?)
- **bitSort(value)** 00101001 ==> 00000111
or with minimal # toggles?
- **bitReverse(uint32_t x, uint8_t n)**
- **byteReverse24(uint32_t x)** dedicated 24 bit = 3 bytes e.g RGB
- **byteInverse(uint32_t x)**  (a,b,c,d) => (255-a, 255-b, 255-c, 255-d) = rather simple ~?
- **isBitPalindrome()** byte, word ...
- **bitSwap(value, p, q)**


#### Functions fix

- **bitRotateLeftRight()** should it do modulo pos?
- **bitRotateLeftRight()** should it handle (pos == 0) separately
- **bitsNeededRef()** correct for value 0?
- **bitRot(value, chance = 50%, times = 1)** extension...
- **bitRot()** one random + mod might be faster (dependency? bad?)


## Operations

See examples.


CHANGELOG.md
# Change Log bitHelpers

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.6] - 2022-04-11

### Added
- add CHANGELOG.md

### Changed
- better masking in **bitCount()**
- added () in several places.
- edit README.md

### Fixed
- Issue #10 Incorrect behaviour in bitWrite64 for value 31
  Specific for ESP8266
- Fix **bitRead64()** return value.


## [0.1.5] - 2021-12-14

### Changed
- update library.json, license, minor edits


## [0.1.4] - 2021-10-19

### Changed 
- update build-CI (e.g. ESP32 compilation of tests)
- update README.md with badges


## [0.1.3] - 2021-08-09

### Changed
- update README.md
- update unit tests


## [0.1.2] - 2020-12-14

### Added
- add Arduino-CI + unit tests


## [0.1.1] - 2020-08-10

### Added
- added BitsNeeded, bitSet64 family


## [0.1.0] - 2020-07-29

### Added
- initial release


## [0.0.1] - 2015-11-07

### Added
- initial version








test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-03
// PURPOSE: unit tests for the bitHelpers.h
//          https://github.com/RobTillaart/bitHelpers
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)

#include <ArduinoUnitTests.h>

#include "bitHelpers.h"


unittest_setup()
{
}

unittest_teardown()
{
}

unittest(test_bitCount)
{
  assertEqual(32, bitCountReference((uint32_t)0xFFFFFFFF));
  assertEqual(32, bitCountKR((uint32_t)0xFFFFFFFF));
  assertEqual(32, bitCountArray((uint32_t)0xFFFFFFFF));
  assertEqual(32, bitCountF1((uint32_t)0xFFFFFFFF));
  assertEqual(32, bitCountF2((uint32_t)0xFFFFFFFF));
  assertEqual(32, bitCount((uint32_t)0xFFFFFFFF));
}

unittest(test_bitReverse)
{
  assertEqual(0xAA, bitReverse((uint8_t)0x55));
  assertEqual(0xFF00, bitReverse((uint16_t)0x00FF));
  assertEqual(0xFFFFAAAA, bitReverse((uint32_t)0x5555FFFF));
  assertEqual(0xFFFFAAAAFFFFAAAA, bitReverse((uint64_t)0x5555FFFF5555FFFF));
}

unittest(test_nybbleReverse)
{
  assertEqual(0xAB, nybbleReverse((uint8_t)0xBA));
  assertEqual(0xABCD, nybbleReverse((uint16_t)0xDCBA));
  assertEqual(0x12345678, nybbleReverse((uint32_t)0x87654321));
  assertEqual(0x0123456789ABCDEF, nybbleReverse((uint64_t)0xFEDCBA9876543210));
}

unittest(test_byteReverse)
{
  assertEqual(0xABCD, byteReverse((uint16_t)0xCDAB));
  assertEqual(0x12345678, byteReverse((uint32_t)0x78563412));
  assertEqual(0x0123456789ABCDEF, byteReverse((uint64_t)0xEFCDAB8967452301));
}

unittest(test_swap)
{
  assertEqual(0x15, swap((uint8_t)0x51));
  assertEqual(0x1234, swap((uint16_t)0x3412));
  assertEqual(0x12345678, swap((uint32_t)0x56781234));
  assertEqual(0x0123456789ABCDEF, swap((uint64_t)0x89ABCDEF01234567));
}

unittest(test_bitRotateLeft)
{
  assertEqual(0x8A, bitRotateLeft((uint8_t)0x51, 3));
  assertEqual(0x4682, bitRotateLeft((uint16_t)0x3412, 13));
  assertEqual(0x1A2B3C09, bitRotateLeft((uint32_t)0x56781234, 23));
  assertEqual(0x456789ABCDEF0123, bitRotateLeft((uint64_t)0x89ABCDEF01234567, 48));
}

unittest(test_bitRotateRight)
{
  assertEqual(0x2A, bitRotateRight((uint8_t)0x51, 3));
  assertEqual(0xA091, bitRotateRight((uint16_t)0x3412, 13));
  assertEqual(0xF02468AC, bitRotateRight((uint32_t)0x56781234, 23));
  assertEqual(0x0123456789ABCDEF, bitRotateRight((uint64_t)0x89ABCDEF01234567, 32));
}

unittest(test_bitFlip)
{
  assertEqual(0x59, bitFlip((uint8_t)0x51, 3));
  assertEqual(0x1412, bitFlip((uint16_t)0x3412, 13));
  assertEqual(0x56F81234, bitFlip((uint32_t)0x56781234, 23));
  assertEqual(0x89ABCD7EF01234567, bitFlip((uint64_t)0x89ABCDEF01234567, 43));
}

unittest(test_bitsNeeded)
{
  assertEqual(15, bitsNeededRef(0x5167));
  assertEqual(6 , bitsNeeded((uint8_t)0x34));
  assertEqual(11, bitsNeeded((uint16_t)0x0412));
  assertEqual(32, bitsNeeded((uint32_t)0xF0001234));
  assertEqual(49, bitsNeeded((uint64_t)0x0001DEF012345677));
}


unittest_main()

// --------


bitHelpers.cpp
//
//    FILE: bitHelpers.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.8
//    DATE: 2015-11-07
// PURPOSE: Arduino library with functions on bit level
//     URL: https://github.com/RobTillaart/bitHelpers
//
// HISTORY: See CHANGELOG.md


#include "bitHelpers.h"


////////////////////////////////////////////////
//
// BIT COUNT TEST
//
uint8_t bitCountReference(uint32_t value)
{
  uint32_t v = value;
  uint8_t count = 0;
  while (v > 0)
  {
    if (v & 1) count++;
    v >>= 1;
  }
  return count;
}


uint8_t bitCountKR(uint32_t value)
{
  // Kerningham & Ritchie
  uint32_t v = value;
  uint8_t count = 0;
  while (v)
  {
    count++;
    v &= (v - 1);
  }
  return count;
}


uint8_t bitCountArray(uint32_t value)
{
  uint8_t ar[] = {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4};
  uint32_t v = value;
  uint8_t count = 0;
  while (v)
  {
    count += ar[v & 0x0F];
    v >>= 4;
  }
  return count;
}


uint8_t bitCountF1(uint32_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint32_t v = value;
  v = v - ((v >> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
  v = (v + (v >> 4)) & 0x0F0F0F0F;
  v = v + (v >> 8);
  v = v + (v >> 16);
  return v & 0x0000003F;
};


uint8_t bitCountF2(uint32_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint32_t v = value;
  v = v - ((v >> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
  v = ((v + (v >> 4)) & 0x0F0F0F0F);
  return ( v * 0x01010101) >> 24;
}


////////////////////////////////////////////////
//
// BIT COUNT
//
uint8_t bitCount(uint8_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint8_t v = value;
  v = v - ((v >> 1) & 0x55);
  v = (v & 0x33) + ((v >> 2) & 0x33);
  v = (v + (v >> 4)) & 0x0F;
  return v;
}


uint8_t bitCount(uint16_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint16_t v = value;
  v = v - ((v >> 1) & 0x5555);
  v = (v & 0x3333) + ((v >> 2) & 0x3333);
  v = (v + (v >> 4)) & 0x0F0F;
  v = (v + (v >> 8)) & 0x1F;
  return v;
}


uint8_t bitCount(uint32_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint32_t v = value;
  v = v - ((v >> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >> 2) & 0x33333333);
  v = (v + (v >> 4)) & 0x0F0F0F0F;
  v =  v + (v >> 8);
  v = (v + (v >> 16)) & 0x3F;
  return v;
}


uint8_t bitCount(uint64_t value)
{
  //  parallel adding in a register SWAG algorithm
  uint64_t v = value;
  v = v - ((v >> 1) & 0x5555555555555555);
  v = (v & 0x3333333333333333) + ((v >> 2) & 0x3333333333333333);
  v = (v + (v >> 4)) & 0x0F0F0F0F0F0F0F0F;
  v =  v + (v >> 8);
  v =  v + (v >> 16);
  v = (v + (v >> 32)) & 0x7F;
  return v;
}


////////////////////////////////////////////////
//
// BIT REVERSE
//
uint8_t bitReverse(uint8_t value)
{
  uint8_t x = value;
  x = (((x & 0xAA) >> 1) | ((x & 0x55) << 1));
  x = (((x & 0xCC) >> 2) | ((x & 0x33) << 2));
  x = (x >> 4) | (x << 4);
  return x;
}


uint16_t bitReverse(uint16_t value)
{
  uint16_t x = value;
  x = (((x & 0xAAAA) >> 1) | ((x & 0x5555) << 1));
  x = (((x & 0xCCCC) >> 2) | ((x & 0x3333) << 2));
  x = (((x & 0xF0F0) >> 4) | ((x & 0x0F0F) << 4));
  x = (x >> 8) | (x << 8);
  return x;
}


uint32_t bitReverse(uint32_t value)
{
  uint32_t x = value;
  x = (((x & 0xAAAAAAAA) >> 1)  | ((x & 0x55555555) << 1));
  x = (((x & 0xCCCCCCCC) >> 2)  | ((x & 0x33333333) << 2));
  x = (((x & 0xF0F0F0F0) >> 4)  | ((x & 0x0F0F0F0F) << 4));
  x = (((x & 0xFF00FF00) >> 8)  | ((x & 0x00FF00FF) << 8));
  x = (x >> 16) | (x << 16);
  return x;
}


uint64_t bitReverse(uint64_t value)
{
  uint64_t x = value;
  x = (((x & 0xAAAAAAAAAAAAAAAA) >> 1)  | ((x & 0x5555555555555555) << 1));
  x = (((x & 0xCCCCCCCCCCCCCCCC) >> 2)  | ((x & 0x3333333333333333) << 2));
  x = (((x & 0xF0F0F0F0F0F0F0F0) >> 4)  | ((x & 0x0F0F0F0F0F0F0F0F) << 4));
  x = (((x & 0xFF00FF00FF00FF00) >> 8)  | ((x & 0x00FF00FF00FF00FF) << 8));
  x = (((x & 0xFFFF0000FFFF0000) >> 16) | ((x & 0x0000FFFF0000FFFF) << 16));
  x = (x >> 32) | (x << 32);
  return x;
}


////////////////////////////////////////////////
//
// NYBBLE REVERSE
//
uint8_t nybbleReverse(uint8_t value)
{
  uint8_t x = value;
  x = (x >> 4) | (x << 4);
  return x;
}


uint16_t nybbleReverse(uint16_t value)
{
  uint16_t x = value;
  x = (((x & 0xF0F0) >> 4) | ((x & 0x0F0F) << 4));
  x = (x >> 8) | (x << 8);
  return x;
}


uint32_t nybbleReverse(uint32_t value)
{
  uint32_t x = value;
  x = (((x & 0xF0F0F0F0) >> 4)  | ((x & 0x0F0F0F0F) << 4));
  x = (((x & 0xFF00FF00) >> 8)  | ((x & 0x00FF00FF) << 8));
  x = (x >> 16) | (x << 16);
  return x;
}


uint64_t nybbleReverse(uint64_t value)
{
  uint64_t x = value;
  x = (((x & 0xF0F0F0F0F0F0F0F0) >> 4)  | ((x & 0x0F0F0F0F0F0F0F0F) << 4));
  x = (((x & 0xFF00FF00FF00FF00) >> 8)  | ((x & 0x00FF00FF00FF00FF) << 8));
  x = (((x & 0xFFFF0000FFFF0000) >> 16) | ((x & 0x0000FFFF0000FFFF) << 16));
  x = (x >> 32) | (x << 32);
  return x;
}


////////////////////////////////////////////////
//
// BYTE REVERSE
//
uint16_t byteReverse(uint16_t value)
{
  uint16_t x = value;
  x = (x >> 8) | (x << 8);
  return x;
}


uint32_t byteReverse(uint32_t value)
{
  uint32_t x = value;
  x = (((x & 0xFF00FF00) >> 8)  | ((x & 0x00FF00FF) << 8));
  x = (x >> 16) | (x << 16);
  return x;
}


uint64_t byteReverse(uint64_t value)
{
  uint64_t x = value;
  x = (((x & 0xFF00FF00FF00FF00) >> 8)  | ((x & 0x00FF00FF00FF00FF) << 8));
  x = (((x & 0xFFFF0000FFFF0000) >> 16) | ((x & 0x0000FFFF0000FFFF) << 16));
  x = (x >> 32) | (x << 32);
  return x;
}


////////////////////////////////////////////////
//
// WORD REVERSE
//
uint32_t wordReverse(uint32_t value)
{
  uint32_t x = value;
  x = (x >> 16) | (x << 16);
  return x;
}


uint64_t wordReverse(uint64_t value)
{
  uint64_t x = value;
  x = (((x & 0xFFFF0000FFFF0000) >> 16) | ((x & 0x0000FFFF0000FFFF) << 16));
  x = (x >> 32) | (x << 32);
  return x;
}


////////////////////////////////////////////////
//
// SWAP HI LO
//
uint8_t swap(uint8_t value)
{
  return (value << 4) | (value >> 4);
}


uint16_t swap(uint16_t value)
{
  return (value << 8) | (value >> 8);
}


uint32_t swap(uint32_t value)
{
  return (value << 16) | (value >> 16);
}


uint64_t swap(uint64_t value)
{
  return (value << 32) | (value >> 32);
}


////////////////////////////////////////////////
//
// BIT ROTATE LEFT
//
uint8_t bitRotateLeft(uint8_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 7) return value;
  return (value << pos) | (value >> (8 - pos));
}


uint16_t bitRotateLeft(uint16_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 15) return value;
  return (value << pos) | (value >> (16 - pos));
}


uint32_t bitRotateLeft(uint32_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 31) return value;
  return (value << pos) | (value >> (32 - pos));
}


uint64_t bitRotateLeft(uint64_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 63) return value;
  return (value << pos) | (value >> (64 - pos));
}


////////////////////////////////////////////////
//
// BIT ROTATE RIGHT
//
uint8_t  bitRotateRight(uint8_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 7) return value;
  return (value << (8 - pos)) | (value >> pos);
}


uint16_t bitRotateRight(uint16_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 15) return value;
  return (value << (16 - pos)) | (value >> pos);
}


uint32_t bitRotateRight(uint32_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 31) return value;
  return (value << (32 - pos)) | (value >> pos);
}


uint64_t bitRotateRight(uint64_t value, uint8_t pos)
{
  if (pos == 0) return value;
  if (pos > 63) return value;
  return (value << (64 - pos)) | (value >> pos);
}


////////////////////////////////////////////////////
//
// BIT FLIP
//
uint8_t bitFlip(uint8_t value, uint8_t pos)
{
  if (pos > 7) return value;
  return value ^ (1 << pos);
}


uint16_t bitFlip(uint16_t value, uint8_t pos)
{
  if (pos > 15) return value;
  return value ^ (1 << pos);
}


uint32_t bitFlip(uint32_t value, uint8_t pos)
{
  if (pos > 31) return value;
  return value ^ (1UL << pos);
}


uint64_t bitFlip(uint64_t value, uint8_t pos)
{
  if (pos > 63) return value;
  return value ^ (1ULL << pos);
}


////////////////////////////////////////////////////
//
// BIT ROT
//
uint8_t bitRotRef(uint8_t value, float chance = 0.5)
{
  if (random(BH_BIG_NR) > chance * BH_BIG_NR) return value;
  return value ^ (1 << random(8));
}


uint16_t bitRotRef(uint16_t value, float chance = 0.5)
{
  if (random(BH_BIG_NR) > chance * BH_BIG_NR) return value;
  return value ^ (1UL << random(16));
}


uint32_t bitRotRef(uint32_t value, float chance = 0.5)
{
  if (random(BH_BIG_NR) > chance * BH_BIG_NR) return value;
  return value ^ (1UL << random(32));
}


uint64_t bitRotRef(uint64_t value, float chance = 0.5)
{
  if (random(BH_BIG_NR) > chance * BH_BIG_NR) return value;
  return value ^ (1ULL << random(64));
}


uint8_t bitRot(uint8_t value, float chance = 0.5, uint16_t times = 1)
{
  while(times--)
  {
    uint32_t r = random(BH_BIG_NR);
    if ( r < chance * BH_BIG_NR)
    {
      value ^= (1 << (r & 7));
    }
  }
  return value;
}


uint16_t bitRot(uint16_t value, float chance = 0.5, uint16_t times = 1)
{
  while(times--)
  {
    uint32_t r = random(BH_BIG_NR);
    if ( r < chance * BH_BIG_NR)
    {
      value ^= (1 << (r & 15));
    }
  }
  return value;
};


uint32_t bitRot(uint32_t value, float chance = 0.5, uint16_t times = 1)
{
  while(times--)
  {
    uint32_t r = random(BH_BIG_NR);
    if ( r < chance * BH_BIG_NR)
    {
      value ^= (1 << (r & 31));
    }
  }
  return value;
};


uint64_t bitRot(uint64_t value, float chance = 0.5, uint16_t times = 1)
{
  while(times--)
  {
    uint32_t r = random(BH_BIG_NR);
    if ( r < chance * BH_BIG_NR)
    {
      value ^= (1 << (r & 63));
    }
  }
  return value;
};




////////////////////////////////////////////////////
//
// BIT-SET64 -CLEAR64 -TOGGLE64 -READ64 -WRITE64
//

// FUNCTIONS
#if defined(__AVR__)
//  optimized for performance

void bitSet64(uint64_t & x, uint8_t n)
{
  if (n > 47)      x |= (0x1000000000000 << (n - 48));
  else if (n > 31) x |= (0x100000000 << (n - 32));
  else if (n > 23) x |= (0x1000000 << (n - 24));
  else if (n > 15) x |= (0x10000 << (n - 16));
  else             x |= (0x1 << n);
}


void bitClear64(uint64_t & x, uint8_t n)
{
  if (n > 47)      x &= ~(0x1000000000000 << (n - 48));
  else if (n > 31) x &= ~(0x100000000 << (n - 32));
  else if (n > 23) x &= ~(0x1000000 << (n - 24));
  else if (n > 15) x &= ~(0x10000 << (n - 16));
  else             x &= ~(0x1 << n);
}


void bitToggle64(uint64_t & x, uint8_t n)
{
  if (n > 47)      x ^= (0x1000000000000 << (n - 48));
  else if (n > 31) x ^= (0x100000000 << (n - 32));
  else if (n > 23) x ^= (0x1000000 << (n - 24));
  else if (n > 15) x ^= (0x10000 << (n - 16));
  else             x ^= (0x1 << n);
}


#elif defined(ESP32)
//  optimized for performance

void bitSet64(uint64_t & x, uint8_t n)
{
  if (n > 31) x |= (0x100000000 << (n - 32));
  else        x |= (0x1 << n);
}


void bitClear64(uint64_t & x, uint8_t n)
{
  if (n > 31) x &= ~(0x100000000 << (n - 32));
  else        x &= ~(0x1 << n);
}


void bitToggle64(uint64_t & x, uint8_t n)
{
  if (n > 31) x ^= (0x100000000 << (n - 32));
  else        x ^= (0x1 << n);
}


#else
//  slower reference implementation

void bitSet64(uint64_t & x, uint8_t bit)
{
  x |= (1ULL << bit);
}


void bitClear64(uint64_t & x, uint8_t bit)
{
  x &= ~(1ULL << bit);
}


void bitToggle64(uint64_t & x, uint8_t bit)
{
  x ^= (1ULL << bit);
}


#endif


uint8_t bitRead64(uint64_t & x, uint8_t bit)
{
  return ((x & (1ULL << bit)) > 0);
}


void bitWrite64(uint64_t & x, uint8_t bit, uint8_t value)
{
  if (value) bitSet64(x, bit);
  else bitClear64(x, bit);
}


////////////////////////////////////////////////////
//
// BITS NEEDED
//

// reference
uint8_t bitsNeededRef(uint64_t x)
{
  uint8_t n = 0;
  while (x)
  {
    x >>= 1;
    n++;
  }
  return n;
}


// workers
uint8_t bitsNeeded(uint8_t x)
{
  uint8_t n = 0;
  while (x)
  {
    x >>= 1;
    n++;
  }
  return n;
}


uint8_t bitsNeeded(uint16_t x)
{
  uint8_t y = x >> 8;
  if (y != 0) return bitsNeeded(y) + 8;
  return bitsNeeded((uint8_t)x);
}


uint8_t bitsNeeded(uint32_t x)
{
  uint16_t y = x >> 16;
  if (y != 0) return bitsNeeded(y) + 16;
  return bitsNeeded((uint16_t)x);
}


uint8_t bitsNeeded(uint64_t x)
{
  uint32_t y = x >> 32;
  if (x >> 32) return bitsNeeded(y) + 32;
  return bitsNeeded((uint32_t)x);
}


// -- END OF FILE --

